server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -309,6 +318,41 @@ public IndexShard(
         primaryTerm = indexSettings.getIndexMetaData().primaryTerm(shardId.id());
         refreshListeners = buildRefreshListeners();
         lastSearcherAccess.set(threadPool.relativeTimeInMillis());
+
+        final TimeValue refreshInterval = indexSettings.getValue(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING);
+        logger.debug("store stats are refreshed with refresh_interval [{}]", refreshInterval);
+        storeStatsCache = new SingleObjectCache<long[]>(refreshInterval, new long[] { -1, -1 }) {
+
+            @Override
+            protected long[] refresh() {
+                try {
+                    return new long[] { store.sizeInBytes(), threadPool.absoluteTimeInMillis() };
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+
+            @Override
+            protected boolean needsRefresh() {
```

Comment:
I am not sure why this is so complex, wouldn't it be enough to override `needsRefresh()`? 

```Java

private MergeStats previousStats = new MergeStats();

if (super.needsRefresh()) {
  boolean refresh = false;
  if (isActive()) {
    MergeStats mergeStats = getEngine().getMergeStats();
    refresh = mergeStats.current > 0 || mergeStats.total != previousStats.total;
    previousStats = mergeStats;
  }
  return refresh;
}
```